chalice/deploy/packager.py

```diff
@@ -190,21 +217,416 @@ def inject_latest_app(self, deployment_package_filename, project_dir):
+        # type: (str) -> Message
+        # The PKG-INFO generated by the egg-info command is in an email feed
+        # format, so we use an email feedparser here to extract the metadata
+        # from the PKG-INFO file.
+        data = self._osutils.get_file_contents(filepath, binary=False)
+        parser = FeedParser()
+        parser.feed(data)
+        return parser.close()
+
+    def _generate_egg_info(self, package_dir):
+        # type: (str) -> str
+        setup_py = self._osutils.joinpath(package_dir, 'setup.py')
+        script = self._SETUPTOOLS_SHIM % setup_py
+
+        cmd = [sys.executable, '-c', script, '--no-user-cfg', 'egg_info']
+        egg_info_dir = self._osutils.joinpath(package_dir, 'egg-info')
+        self._osutils.makedirs(egg_info_dir)
+        cmd += ['--egg-base', 'egg-info']
+        p = subprocess.Popen(cmd, cwd=package_dir,
+                             stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        p.communicate()
+        self._osutils.joinpath(egg_info_dir)
+        info_contents = self._osutils.get_directory_contents(egg_info_dir)
+        assert len(info_contents) == 1
+        pkg_info_path = self._osutils.joinpath(
+            egg_info_dir, info_contents[0], 'PKG-INFO')
+        return pkg_info_path
+
+    def _unpack_sdist_into_dir(self, sdist_path, unpack_dir):
+        # type: (str, str) -> str
+        if sdist_path.endswith('.zip'):
+            self._osutils.unpack_zipfile(sdist_path, unpack_dir)
+        elif sdist_path.endswith('.tar.gz'):
+            self._osutils.unpack_tarfile(sdist_path, unpack_dir)
+        else:
+            raise InvalidSourceDistributionNameError(sdist_path)
+        # There should only be one directory unpacked.
+        contents = self._osutils.get_directory_contents(unpack_dir)
+        assert len(contents) == 1
+        return self._osutils.joinpath(unpack_dir, contents[0])
+
+    def get_package_name_and_version(self, sdist_path):
+        # type: (str) -> Tuple[str, str]
+        with self._osutils.tempdir() as tempdir:
+            package_dir = self._unpack_sdist_into_dir(sdist_path, tempdir)
+            pkg_info_filepath = self._generate_egg_info(package_dir)
+            metadata = self._parse_pkg_info_file(pkg_info_filepath)
+            name = metadata['Name']
+            version = metadata['Version']
+        return name, version
+
+
+class SubprocessPip(object):
+    """Wrapper around calling pip through a subprocess."""
+    def main(self, args):
+        # type: (List[str]) -> Tuple[int, Optional[bytes]]
+        python_exe = sys.executable
+        invoke_pip = [python_exe, '-m', 'pip']
+        invoke_pip.extend(args)
+        p = subprocess.Popen(invoke_pip,
+                             stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        _, err = p.communicate()
+        rc = p.returncode
+        return rc, err
+
+
+class PipRunner(object):
+    """Wrapper around pip calls used by chalice."""
+
```

Comment:
Sorry to nitpick, but this is the only place you've got a space between the class doc string and the first method.